File: ts/2022/day21.ts
import { main } from 'lib/advent';
import { lines } from 'lib/util';

function parse(lines: string[]) {
  return new Map<string, string | number>(
    lines.map((line) => {
      const [name, expr] = line.split(': ');
      return [name, isNaN(+expr) ? expr : Number(expr)];
    })
  );
}

const ops: Record<string, (a: number, b: number) => number> = {
  '+': (a, b) => a + b,
  '-': (a, b) => a - b,
  '*': (a, b) => a * b,
  '/': (a, b) => a / b,
};

function resolveMonkey(
  monkeys: Map<string, string | number>,
  start = 'root'
): number {
  const val = monkeys.get(start);
  if (typeof val === 'number') return val;
  const [a, opStr, b] = val.split(' ');
  const op = ops[opStr];
  return op(resolveMonkey(monkeys, a), resolveMonkey(monkeys, b));
}

function makeString(
  monkeys: Map<string, string | number>,
  name: string
): string {
  if (name === 'humn') return name;
  const val = monkeys.get(name);
  if (typeof val === 'number') return val.toString();

  const [a, opStr, b] = val.split(' ');
  const aVal = makeString(monkeys, a);
  const bVal = makeString(monkeys, b);
  if (!isNaN(+aVal) && !isNaN(+bVal)) {
    return ops[opStr](Number(aVal), Number(bVal)).toString();
  }
  return `(${aVal} ${opStr} ${bVal})`;
}

// TODO: clean up
function part2(monkeys: Map<string, string | number>, start = 'root'): number {
  const val = monkeys.get(start);
  if (typeof val === 'number') throw new Error();
  const [targetA, , targetB] = val.split(' ');

  let test = makeString(monkeys, targetA);
  let goal = makeString(monkeys, targetB);
  if (isNaN(+goal)) [test, goal] = [goal, test];
  const goalVal = Number(eval(goal));

  let lowBound = Number.MIN_SAFE_INTEGER;
  let highBound = Number.MAX_SAFE_INTEGER;

  // TODO: reuse part1 instead of eval?
  const checkPos = eval(test.replace('humn', '1'));
  const checkNeg = eval(test.replace('humn', '-1'));
  const invert = checkPos < checkNeg;
  const isTooHigh = (check: number): boolean =>
    invert ? check < goalVal : check > goalVal;
  for (;;) {
    const humn = Math.floor((highBound + lowBound) / 2);
    const check = eval(test);
    if (check === goalVal) return humn;
    if (isTooHigh(check)) highBound = humn;
    else lowBound = humn;
  }

  return 0;
}

main(
  (s) => resolveMonkey(parse(lines(s))),
  (s) => part2(parse(lines(s)))
);

File: py/2022/day21.py
import operator
import sys

from aoc import main
from aoc.parse import line_parser

OPS = {
    "+": operator.add,
    "-": operator.sub,
    "*": operator.mul,
    "/": operator.truediv,
}


@line_parser
def parse(line: str) -> tuple[str, int | str]:
    name, expr = line.split(": ")
    return name, (int(expr) if expr.isdigit() else expr)


def yell(name: str, monkeys: dict[str, int | str]) -> int:
    val = monkeys[name]
    if isinstance(val, int):
        return val
    left, op, right = val.split(" ")
    return OPS[op](yell(left, monkeys), yell(right, monkeys))


def human_yell(monkeys: dict[str, int | str]) -> int:
    assert isinstance(monkeys["root"], str)
    a, _, b = monkeys["root"].split(" ")
    zero = yell(a, dict(monkeys, humn=0))
    one = yell(a, dict(monkeys, humn=1))
    op = operator.gt if one > zero else operator.lt
    lo = -sys.maxsize
    hi = sys.maxsize
    while True:
        humn = (hi + lo) // 2
        val_a = yell(a, dict(monkeys, humn=humn))
        val_b = yell(b, dict(monkeys, humn=humn))
        if val_a == val_b:
            return humn
        elif op(val_a, val_b):
            hi = humn
        else:
            lo = humn


if __name__ == "__main__":
    main(
        lambda s: int(yell("root", dict(parse(s)))),
        lambda s: human_yell(dict(parse(s))),
    )
